### DISCUSSION TITLE
Gdb support for picolibc TLS

### DISCUSSION THREAD

**Discussion by hoxi** [picolibc/picolibc#745] [Q&A]
We have replaced newlib with picolibc in our project, and our first step has been to enable TLS together with FreeRTOS. Following the documentation, it all seem to work very well (apart from that some IDE FreeRTOS integrations was not prepared for finding TLS variables on the stack it seems).

One issue is that TLS variables does not seem to be understood by GDB (this example is from a target debug session):

```
gdb) print my_tls_var
Cannot find thread-local storage for Thread 57005, executable file test.axf:
TLS not supported or disabled on this target
(gdb) show remote get-thread-local-storage-address
Support for the 'qGetTLSAddr' packet on the current remote target is "auto", currently disabled.
```

We are using our own built GDB and target a ARM Cortex-M7.

Are we missing any configuration in GDB, or is there any additional support in picolibc needed for GDB to query libc for TLS storage?

**Comment by keith-packard** [comment]
I've never managed to get this working either. Reading through the gdb and qemu source code, I'm not sure there's any way to configure yourself to success here. Somehow you need to get the debugger to read the location of the TLS base (either from the __tls variable or one of the target-specific registers) and use that with the offset of the variable to find the address in memory.

**Comment by hoxi** [comment]
Following up on this — we've implemented a patch against GDB 17.1 that adds TLS resolution for bare-metal/RTOS ARM targets using picolibc. It handles both the hardware register path (Cortex-A/R where picolibc writes TPIDRURO via CP15) and the __tls global variable path (Cortex-M where no hardware thread pointer exists).

[arm-none-tls-picolibc.patch](https://github.com/user-attachments/files/26241565/arm-none-tls-picolibc.patch)

Not sure about a proper detection logic, but we currently use two methods to identify bare-metal binaries in the ARM OSABI sniffer:

- **Explicit**: A .note.abinone ELF note section that the binary can opt into (via a small assembly file or linker script). If present, GDB immediately selects GDB_OSABI_NONE.
- **Heuristic**: If the ELF is an executable with no .interp, no .dynamic, and no .note.ABI-tag, treat it as bare-metal. The .note.ABI-tag check excludes statically linked Linux binaries since glibc's startup objects always emit that note.

For the TLS resolution itself, we validate picolibc by checking that both __tls and _set_tls symbols exist before using the __tls variable. On Cortex-A/R, the hardware TLS register is tried first via the org.gnu.gdb.arm.tls target description feature.

Not sure if the GDB folks would be open for this type of patch, but I might try submitting it upstream after more thorough testing. Would be interested to hear if anyone has thoughts on the detection approach or wants to try the patch on other ARM profiles.

**Comment by keith-packard** [comment]
I've been wanting to figure this out for years now, thanks so much. I wonder whether this also works with Zephyr, which provides TLS support using a mechanism similar to picolibc but implemented independently.

As for getting things upstream, I'd definitely encourage that; it can be a bit of a challenge, but it's well worth the effort as this will then flow into every toolchain instead of needing to be picked up and added separately. Zephyr provides toolchains including gdb for a wide range of targets; once you're reasonably happy with how this code works, I can help get changes integrated and delivered there.

**Comment by keith-packard** [comment]
I've been playing with this a bit today on top of GDB 16 and it worked great for targets using `__tls`, but targets using the MSR never get to `create_feature_arm_arm_tls` because `arm-none-tdep.c` doesn't have any of the extra register parsing logic present in the other arm targets. Am I missing something?

**Comment by hoxi** [comment]
Thanks for trying this! The register parsing for `org.gnu.gdb.arm.tls` is actually in `arm-tdep.c` (in `arm_gdbarch_init`). But it depends entirely on the GDB server advertising the TLS feature in its target description XML. I guess the problem is that very few (if any?) GDB servers have included this feature register in their target descriptions. We use both J-Link GDB Server and OpenOCD, and from what we could see none of them currently include `org.gnu.gdb.arm.tls` in their target descriptions for Cortex-A/R. So we simply added this manually to get it working. I did not want to add any code to GDB that tried to probe that register, that can probably have unwanted side effects depending on hardware.

Maybe you can try to grab the target description from your GDB server and add the TLS feature to it:

1. Connect to the target normally and dump the current description: 
 `(gdb) maint print xml-tdesc`
   
2. Add this feature register:
   ```xml
   <feature name="org.gnu.gdb.arm.tls">
     <reg name="tpidruro" bitsize="32" type="data_ptr"/>
   </feature>
   ```
And then use this target description instead.

Looked a bit at Zephyr and I agree that we would want a solution that could work out of the box with both vanilla picolibc and Zephyr. Unfortunately they seem to use `z_arm_tls_ptr` as the TLS pointer symbol. And they do not define `__arm32_tls_tcb_offset`, but instead `z_arm_tls_ptr` points to the current tls block directly from what I can understand. It would have been nice to align around a `__tls` symbol as some kind of ABI (with optional use of `__arm32_tls_tcb_offset` according to ARM ABI). Probably easier to convince the GDB folks as well if it is nice and clean. But that would also mean that picolibc would have to add `__attribute__((weak))` for `__tls`, and Zephyr renames `z_arm_tls_ptr` to `__tls` (seems to be a small patch).

So I guess we are looking at (somewhat in order):
- **GDB**: Upstream the GDB patch, defining `__tls` as the convention to point to the TLS block if the target description does not provide `org.gnu.gdb.arm.tls`
- **Picolibc**: Make `__tls` a weak symbol, allowing RTOS (or other implementation) override of `__tls`
- **Zephyr**: Try to upstream `z_arm_tls_ptr` -> `__tls`, citing that this will allow Zephyr TLS to work with GDB out of the box (assuming we are successful with GDB)
- **OpenOCD**: Try to upstream `org.gnu.gdb.arm.tls` in their target definitions for relevant architectures. If I'm not mistaken this was defined as a feature already in GDB 13.

What do you think? In my head this would be a nice way forward, but maybe I'm just simplifying this too much😄.

**Reply by keith-packard** [reply]
> Thanks for trying this! The register parsing for `org.gnu.gdb.arm.tls` is actually in `arm-tdep.c` (in `arm_gdbarch_init`). But it depends entirely on the GDB server advertising the TLS feature in its target description XML. I guess the problem is that very few (if any?) GDB servers have included this feature register in their target descriptions. We use both J-Link GDB Server and OpenOCD, and from what we could see none of them currently include `org.gnu.gdb.arm.tls` in their target descriptions for Cortex-A/R. So we simply added this manually to get it working. I did not want to add any code to GDB that tried to probe that register, that can probably have unwanted side effects depending on hardware.
> 
> Maybe you can try to grab the target description from your GDB server and add the TLS feature to it:
> 
>     1. Connect to the target normally and dump the current description:
>        `(gdb) maint print xml-tdesc`
> 
>     2. Add this feature register:
>        ```
>        <feature name="org.gnu.gdb.arm.tls">
>          <reg name="tpidruro" bitsize="32" type="data_ptr"/>
>        </feature>
>        ```
> 
> 
> And then use this target description instead.
> 
> Looked a bit at Zephyr and I agree that we would want a solution that could work out of the box with both vanilla picolibc and Zephyr. Unfortunately they seem to use `z_arm_tls_ptr` as the TLS pointer symbol. And they do not define `__arm32_tls_tcb_offset`, but instead `z_arm_tls_ptr` points to the current tls block directly from what I can understand. It would have been nice to align around a `__tls` symbol as some kind of ABI (with optional use of `__arm32_tls_tcb_offset` according to ARM ABI). Probably easier to convince the GDB folks as well if it is nice and clean. But that would also mean that picolibc would have to add `__attribute__((weak))` for `__tls`, and Zephyr renames `z_arm_tls_ptr` to `__tls` (seems to be a small patch).
> 
> So I guess we are looking at (somewhat in order):
> 
>     * **GDB**: Upstream the GDB patch, defining `__tls` as the convention to point to the TLS block if the target description does not provide `org.gnu.gdb.arm.tls`

How does `create_feature_arm_arm_tls` work in gdb? Does it only enable use of target descriptions that include the `tpidruro` register or does it add the register based upon gdb internal knowledge? It would be really nice if we could make this work without changing all of the gdb stubs, although if that's required, we can definitely address that. Getting qemu working would make it very easy to debug other bits.

>     * **Picolibc**: Make `__tls` a weak symbol, allowing RTOS (or other implementation) override of `__tls`

I don't think that's necessary or helpful; it's entirely a connection between the TLS setup code and `__aeabi_read_tp`. Picolibc provides an implementation of both ends, but Zephyr provides it's own alternative version. The only requirement is that the desired implementation get linked in first.

>     * **Zephyr**: Try to upstream `z_arm_tls_ptr` -> `__tls`, citing that this will allow Zephyr TLS to work with GDB out of the box (assuming we are successful with GDB)

I can manage that easily enough.

>     * **OpenOCD**: Try to upstream `org.gnu.gdb.arm.tls` in their target definitions for relevant architectures. If I'm not mistaken this was defined as a feature already in GDB 13.

That does seem useful in general; is there any way to do this *and* make gdb have some internal mechanism?

> What do you think? In my head this would be a nice way forward, but maybe I'm just simplifying this too much😄.

I think the only piece you've missed that I care about is qemu. If there are other important gdb stub implementations, we'd want to include those too.
